packages/react-server/src/ReactFlightServer.js

```diff
@@ -42,77 +44,241 @@ type ReactModelObject = {
+      }
     }
   }
+
   return value;
 }
 
+function emitErrorChunk(
+  request: OpaqueRequest,
+  id: number,
+  error: mixed,
+): void {
+  // TODO: We should not leak error messages to the client.
+  // Give this an error code instead and log on the server.
+  let errorMessage;
+  try {
+    errorMessage = '' + (error: any);
+  } catch (x) {
+    errorMessage =
+      'An error occurred but serializing the error message failed.';
+  }
+  let row = serializeRowHeader('E', id) + errorMessage + '\n';
+  request.completedErrorChunks.push(convertStringToBuffer(row));
+}
+
+function retrySegment(request: OpaqueRequest, segment: Segment): void {
+  let value = segment.model;
+  try {
+    while (
+      typeof value === 'object' &&
+      value !== null &&
+      value.$$typeof === REACT_ELEMENT_TYPE
+    ) {
+      // If this is a nested model, there's no need to create another chunk,
+      // we can reuse the existing one and try again.
+      let element: React$Element<any> = (value: any);
+      segment.model = element;
+      value = attemptResolveModelComponent(element);
+    }
+    let json = stringify(value, request.toJSON);
+    let row;
+    let id = segment.id;
+    if (id === 0) {
+      row = json + '\n';
+    } else {
+      row = serializeRowHeader('J', id) + json + '\n';
+    }
+    request.completedJSONChunks.push(convertStringToBuffer(row));
+  } catch (x) {
+    if (typeof x === 'object' && x !== null && typeof x.then === 'function') {
+      // Something suspended again, let's pick it back up later.
+      let ping = segment.ping;
+      x.then(ping, ping);
+      return;
+    } else {
+      // This errored, we need to serialize this error to the
+      emitErrorChunk(request, segment.id, x);
+    }
+  }
+}
+
 function performWork(request: OpaqueRequest): void {
-  let rootModel = request.model;
-  request.model = null;
-  let json = JSON.stringify(rootModel, resolveModelToJSON);
-  request.completedChunks.push(convertStringToBuffer(json));
+  let pingedSegments = request.pingedSegments;
+  request.pingedSegments = [];
+  for (let i = 0; i < pingedSegments.length; i++) {
+    let segment = pingedSegments[i];
+    retrySegment(request, segment);
+  }
   if (request.flowing) {
     flushCompletedChunks(request);
   }
-
-  flushBuffered(request.destination);
 }
 
-function flushCompletedChunks(request: OpaqueRequest) {
+function flushCompletedChunks(request: OpaqueRequest): void {
   let destination = request.destination;
-  let chunks = request.completedChunks;
-  request.completedChunks = [];
-
   beginWriting(destination);
   try {
-    for (let i = 0; i < chunks.length; i++) {
-      let chunk = chunks[i];
-      writeChunk(destination, chunk);
+    let jsonChunks = request.completedJSONChunks;
+    let i = 0;
+    for (; i < jsonChunks.length; i++) {
+      request.pendingChunks--;
+      let chunk = jsonChunks[i];
+      if (!writeChunk(destination, chunk)) {
+        request.flowing = false;
```

Comment:
I don't understand what this means and what the `flowing` property is doing. Can you explain a bit?

Reply:
Flowing means that we can keep writing more data to the buffer once we have it.

We don’t write eagerly to the output stream if it is full. That lets us emit higher pri things if they become available before the stream buffer becomes not full.